File: C++/섹션05. 완전 탐색 기본/03-01. BOJ 2606 - 바이러스 (BFS, 인접 행렬).cpp
#include <stdio.h>

#define MAX (100 + 10)

int V, E; // Vertex, Edge
int MAP[MAX][MAX];

int queue[MAX];
bool visit[MAX];

void input()
{
	scanf("%d %d", &V, &E);

	for (int i = 0; i < E; i++)
	{
		int n1, n2;

		scanf("%d %d", &n1, &n2);

		MAP[n1][n2] = MAP[n2][n1] = 1;
	}
}

void printMap() // for debug
{
	for (int r = 1; r <= V; r++)
	{
		for (int c = 1; c <= V; c++)
			printf("%d ", MAP[r][c]);
		putchar('\n');
	}
	putchar('\n');
}

void BFS(int node)
{
	int rp, wp, count;

	rp = wp = 0;

	queue[wp++] = node;

	visit[node] = true;

	count = 0;
	while (rp < wp)
	{
		int out = queue[rp++];

		for (int c = 1; c <= V; c++)
		{
			if (MAP[out][c] == 0 || visit[c] == true) continue;

			count++;

			queue[wp++] = c;

			visit[c] = true;
		}
	}

	printf("%d\n", count);
}

int main()
{
	input();

	BFS(1);

	return 0;
}

File: Python/섹션05. 완전 탐색 기본/03-01. BOJ 2606 - 바이러스 (BFS, 인접 행렬).py
MAX = 100 + 10

MAP = [[0] * MAX for _ in range(MAX)]

queue = [0] * MAX
visit = [False] * MAX

def input_data():
    global V, E
 
    V = int(input()) # Vertex
    E = int(input()) # Edge

    for _ in range(E):
        n1, n2 = map(int, input().split())     
        MAP[n1][n2] = MAP[n2][n1] = 1
        
def print_map(): # for debug
    for r in range(1, V + 1):
        print(*MAP[r][1:V + 1])

def BFS(node):
    rp, wp = 0, 0

    queue[wp] = node
    wp += 1
    
    visit[node] = True
    
    count = 0
    while rp < wp:
        out = queue[rp]
        rp += 1
        
        for c in range(1, V + 1):
            if MAP[out][c] == 0 or visit[c] == True: continue
                        
            count += 1
            
            queue[wp] = c
            wp += 1
            
            visit[c] = True
        
    print(count)
    
input_data()
        
BFS(1)

